cms/cms_toolbars.py

```diff
@@ -86,6 +88,107 @@ def add_wizard_button(self):
                                       disabled=disabled,
                                       on_close=REFRESH_PAGE)
 
+    def _render_object_editable_buttons(self):
+        self.init_placeholders()
+
+        if not self.toolbar.obj:
+            return
+
+        # Add edit button when in content mode.
+        if self.toolbar.content_mode_active:
+            self.add_edit_button()
+        # Add preview button when in edit mode.
+        if self.toolbar.edit_mode_active:
+            self.add_preview_button()
+        # Always add structure mode.
+        self.add_structure_mode()
+
+    def init_placeholders(self):
+        request = self.request
+        toolbar = self.toolbar
+
+        if toolbar._async and 'placeholders[]' in request.GET:
+            # AJAX request to reload page structure.
+            placeholder_ids = request.GET.getlist('placeholders[]')
+            self.placeholders = Placeholder.objects.filter(pk__in=placeholder_ids)
+        else:
+            if toolbar.structure_mode_active and not toolbar.uses_legacy_structure_mode:
+                # User has explicitly requested structure mode.
+                # and the object (page, blog, etc..) allows for the non-legacy structure mode.
+                renderer = toolbar.structure_renderer
+            else:
+                renderer = toolbar.get_content_renderer()
+
+            self.placeholders = renderer.get_rendered_placeholders()
+
+    def add_edit_button(self):
+        if self.page and not user_can_change_page(self.request.user, page=self.page):
+            return
+
+        item = ButtonList(side=self.toolbar.RIGHT)
+        url = get_object_edit_url(self.toolbar.obj, language=self.toolbar.request_language)
+        item.add_button(
+            _('Edit'),
+            url=url,
+            disabled=False,
+            extra_classes=['cms-btn', 'cms-btn-action', 'cms-btn-switch-edit'],
+        )
+        self.toolbar.add_item(item)
+
+    def add_preview_button(self):
+        if self.page and not user_can_change_page(self.request.user, page=self.page):
```

Comment:
do the check in the caller instead or create a hidden method like _can_add_preview_button and then call the add_preview_button if true. Basically decouple the check from the action.